scale/utils.py

```diff
@@ -0,0 +1,130 @@
+#/usr/bin/python3
+
+"""Kubernetes API access functions"""
+
+from kubernetes import client, config
+#import requests
+import logging
+
+scale_logger = logging.getLogger("scale")
+logging.getLogger("kubernetes").setLevel(logging.WARNING)
+config.load_kube_config()
+v1 = client.CoreV1Api()
+
+
+# def generate_url(host, port):
+#    return "http://" + host + ':' + port + "/api/v1/"
+
+
+def get_nodes():
+    """Return a list of v1.Node"""
+    scale_logger.debug("Getting all nodes in the cluster")
+    return v1.list_node().items
+#     r = requests.get(generate_url(API_HOST, API_PORT) +
+#                      "nodes")
+#     assert r.status_code == 200
+#     try:
+#         nodesList = json.loads(r.text)
+#         return nodesList['items']
+#     except Exception as e:
+#         # FIXME: proper exception handling
+#         print(str(e))
+#         sys.exit(1)
+
+
+def get_pods():
+    """Return a list of v1.Pod"""
+    scale_logger.debug("Getting all pods in all namespaces")
+    return v1.list_pod_for_all_namespaces().items
+#     r = requests.get(generate_url(API_HOST, API_PORT) +
+#                      "pods")
+#     assert r.status_code == 200
+#     try:
+#         podsList = json.loads(r.text)
+#         return podsList['items']
+#     except Exception as e:
+#         # FIXME: proper exception handling
+#         print(str(e))
+#         sys.exit(1)
+
+
+# def get_namespaces_name():
+#     """Return a list of namespaces in the form of string"""
+#     r = requests.get(generate_url(API_HOST, API_PORT) +
+#                      "namespaces")
+#     assert r.status_code == 200
+#     result = []
+#     try:
+#         namespaces = json.loads(r.text)
+#         for each in namespaces.items:
+#             result.append(get_name(each))
+#         return result
+#     except Exception as e:
+#         # FIXME: proper exception handling
+#         print(str(e))
+#         sys.exit(1)
+
+
+def get_pod_host_name(pod):
+    """Return the host node name of the pod"""
+    # Based on Kubernetes API:
+    # https://kubernetes.io/docs/api-reference/v1/definitions/#_v1_podspec
+    # ** API is unclear the value of nodeName flag after the pod is scheduled
+    return pod.spec.node_name
+
+
+def get_cluster_name(node=get_nodes()[0]):
+    """Return the (guessed) name of the cluster"""
+    node_name = get_name(node)
+    parts = node_name.split('-')
+    assert len(parts) > 2
+    return parts[1]
+
+
+def get_pod_name(pod):
+    """Return the name of the pod"""
+    return get_name(pod)
+
+
+def get_node_name(node):
+    """Return the name of the node"""
+    return get_name(node)
+
+
+def get_pod_namespace(pod):
+    """Return the namespace of the pod"""
+    return pod.metadata.namespace
+
+
+def get_pod_type(pod):
+    """Return the Type of the pod"""
+    # TODO: May not be the best approach
+    return get_pod_name(pod).split('-')[0]
```

Comment:
`ownerRefence` is documented here: https://kubernetes.io/docs/api-reference/autoscaling/v1/definitions/#_v1_objectmeta

However, this field does not seem to show up when I made api calls. Here is an example metadata:

"metadata": {
    "name": "gke-dev-default-pool-dbd2a02e-t1lf",
    "selfLink": "/api/v1/nodesgke-dev-default-pool-dbd2a02e-t1lf",
    "uid": "a34c3698-f71c-11e6-ade6-42010a80001a",
    "resourceVersion": "3819560",
    "creationTimestamp": "2017-02-20T03:28:25Z",
    "labels": {
      "beta.kubernetes.io/arch": "amd64",
      "beta.kubernetes.io/instance-type": "n1-standard-4",
      "beta.kubernetes.io/os": "linux",
      "cloud.google.com/gke-nodepool": "default-pool",
      "failure-domain.beta.kubernetes.io/region": "us-central1",
      "failure-domain.beta.kubernetes.io/zone": "us-central1-a",
      "kubernetes.io/hostname": "gke-dev-default-pool-dbd2a02e-t1lf"
    },
    "annotations": {
      "volumes.kubernetes.io/controller-managed-attach-detach": "true"
    }
  }